reactos/drivers/usb/usbohci_new/roothub.c

```diff
@@ -0,0 +1,552 @@
+#include "usbohci.h"
+
+#define NDEBUG
+#include <debug.h>
+
+OHCI_REG_RH_DESCRIPTORA
+NTAPI 
+OHCI_ReadRhDescriptorA(IN POHCI_EXTENSION OhciExtension)
+{
+    POHCI_OPERATIONAL_REGISTERS OperationalRegs;
+    OHCI_REG_RH_DESCRIPTORA DescriptorA;
+    PULONG DescriptorAReg;
+    ULONG ix;
+
+    OperationalRegs = OhciExtension->OperationalRegs;
+    DescriptorAReg = (PULONG)&OperationalRegs->HcRhDescriptorA;
+
+    DPRINT("OHCI_ReadRhDescriptorA: OhciExtension - %p\n", OhciExtension);
+
+    for (ix = 0; ix < 10; ix++)
+    {
+        DescriptorA.AsULONG = READ_REGISTER_ULONG(DescriptorAReg);
+
+        if (DescriptorA.AsULONG != 0 &&
+            DescriptorA.Reserved == 0 &&
+            DescriptorA.NumberDownstreamPorts <= OHCI_MAX_PORT_COUNT) // Reserved bits
+        {
+            break;
+        }
+
+        DPRINT1("OHCI_ReadRhDescriptorA: DescriptorA - %lX, ix - %d\n",
+                DescriptorA.AsULONG, ix);
+
+        KeStallExecutionProcessor(5);
+    }
+
+    return DescriptorA;
+}
+
+VOID
+NTAPI
+OHCI_RH_GetRootHubData(IN PVOID ohciExtension,
+                       IN PVOID rootHubData)
+{
+    POHCI_EXTENSION OhciExtension;
+    PUSBPORT_ROOT_HUB_DATA RootHubData;
+    OHCI_REG_RH_DESCRIPTORA DescriptorA;
+    UCHAR PowerOnToPowerGoodTime;
+    USBPORT_HUB_11_CHARACTERISTICS HubCharacteristics;
+
+    OhciExtension = ohciExtension;
+
+    DPRINT("OHCI_RH_GetRootHubData: OhciExtension - %p, rootHubData - %p\n",
+           OhciExtension,
+           rootHubData);
+
+    RootHubData = rootHubData;
+    DescriptorA = OHCI_ReadRhDescriptorA(OhciExtension);
+
+    RootHubData->NumberOfPorts = DescriptorA.NumberDownstreamPorts;
+
+    /* Waiting time (in 2 ms intervals) */
+    PowerOnToPowerGoodTime = DescriptorA.PowerOnToPowerGoodTime;
+    if (PowerOnToPowerGoodTime <= OHCI_MINIMAL_POTPGT)
+    {
+        PowerOnToPowerGoodTime = OHCI_MINIMAL_POTPGT;
+    }
+    RootHubData->PowerOnToPowerGood = PowerOnToPowerGoodTime;
+
+    HubCharacteristics.AsUSHORT = 0;
+
+    if (DescriptorA.PowerSwitchingMode)
+    {
+        /* Individual port power switching */
+        HubCharacteristics.PowerControlMode = 1;
+    }
+    else
+    {
+        /* Ganged power switching */
+        HubCharacteristics.PowerControlMode = 0;
+    }
+
+    HubCharacteristics.NoPowerSwitching = 0;
+
+    /* always 0 (OHCI RH is not a compound device) */
+    ASSERT(DescriptorA.DeviceType == 0);
```

Comment:
See  7.4.1  HcRhDescriptorARegister, Figure 7-19: HcRhDescriptorARegister
Yes, only two values, since this is 1 bit. I took the name from the specification.

Reply:
Yeah okay looking at the spec, this makes sense. That field seems rather pointless ;)